### ISSUE TITLE
Get menu payload

### ISSUE THREAD

**Issue by dauletbaev** [grammyjs/menu#20]
How can I get menu payload?

```js
feature.on('callback_query:data', async (ctx, next) => {
   ctx.answerCallbackQuery();

   const payload = ctx.callbackQuery.data; // mymenu/0/0/mypayload/{hash}
}
```

How can I get exact payload or have we some method `ctx.getMenuPayload()`?

**Comment by KnorpelSenf** [comment]
Check out https://grammy.dev/plugins/menu.html#payloads. The second sentence says that the payload will be available as `ctx.match`.

**Comment by dauletbaev** [comment]
> Check out https://grammy.dev/plugins/menu.html#payloads. The second sentence says that the payload will be available as `ctx.match`.

```js
// These values copied from console
console.log(ctx.callbackQuery.data); // driver_regions/a/0/tashkent_region/hUK+i
console.log(ctx.match); // undefined
```

`ctx.match` is undefined whenever I tried.

**Comment by KnorpelSenf** [comment]
> `ctx.match` is undefined whenever I tried.

Right, this only works in menu handlers. The menu is meant to handle its own callback queries, and it will provide you with all the information that exists. So there should not really be a reason to need access to it outside of the menu.

If you still must do it, you can of course just extract the payload from the callback query data. This is undocumented behaviour and it may be subject to breaking changes without further notice, so I don't think it is a good idea to do it—but right now, this would work:

```ts
const [,,,payload] = ctx.callbackQuery.data.split('/')
console.log(payload)
```

**Comment by KnorpelSenf** [comment]
It is a very elegant solution to always to through the menu instance. That way, everything will be well encapsulated and abstracted away.

Why do you need to access the menu payload in a location in your code that is unrelated to the menu?

**Comment by dauletbaev** [comment]
> Why do you need to access the menu payload in a location in your code that is unrelated to the menu?

So, I am creating menu through looping over array (I don't want to write every button handler in the loop). Then I register menu in my composer there I am listening callbackQueries.

**Comment by KnorpelSenf** [comment]
Having the handlers is the main point of this plugin. Without handlers, this plugin is useless. So if you do not use them anyway, it may be much simpler to just use the buit-in plugin for creating static inline keyboards. See this: https://grammy.dev/plugins/keyboard.html#inline-keyboards

**Comment by dauletbaev** [comment]
> Having the handlers is the main point of this plugin. Without handlers, this plugin is useless. So if you do not use them anyway, it may be much simpler to just use the buit-in plugin for creating static inline keyboards. See this: https://grammy.dev/plugins/keyboard.html#inline-keyboards

Yeah it's something I want but I am using i18n so I have to use context in order to get translate function. I would like to use static payload if it is possible.

**Comment by KnorpelSenf** [comment]
Oh so you are using dynamic labels? Yeah then I guess the menu plugin works better.

Is it an option to define your handler outside of your loop, and then just pass the same function when creating the buttons? That way, you will be able to access the payload without having to clutter your loop.

**Comment by dauletbaev** [comment]
Thanks I'll try it.

**Comment by KnorpelSenf** [comment]
It's actually interesting that the payload is not there. I would expect that the implementation correctly sets `ctx.match` also outside handlers: https://github.com/grammyjs/menu/blob/8d5c61d020f3cf4e87c5ac3876bd9f3c482ee532/src/menu.ts#L860

**Comment by dauletbaev** [comment]
> It's actually interesting that the payload is not there. I would expect that the implementation correctly sets `ctx.match` also outside handlers:
> 
> https://github.com/grammyjs/menu/blob/8d5c61d020f3cf4e87c5ac3876bd9f3c482ee532/src/menu.ts#L860

I tried to reproduce this behaviour but I had not saved that implementation I have wrote. Whenever I get this issue I will comment here.

**Comment by KnorpelSenf** [comment]
Thanks!

**Comment by KnorpelSenf** [comment]
I will close this for now. Feel free to reopen or comment out create a new issue.

**Comment by MitkaZ7** [comment]
hi, Can anybody help with this plugin?
My menu with multiple submenus  doesn't have any payloads.
First level has 1-2 buttons, second level has multiple btns, and third level has final buttons, and i need to catch callback data at each level, but  i catch it at only last level of menu? why? i can catch data without this plugin, via built-in InlineKeyboard only.

```
 // first level of menu
const menu = new Menu('start-menu')
  .submenu({ text: 'Price', payload: 'price'}, 'rate-menu');

// second level of menu
export const rateMenu = new Menu('rate-menu')
  .submenu({text:'prof', payload: 'prof'}, 'period-menu').submenu('base', 'period-menu')
  .submenu('1C','license-menu').row()
  .submenu('fresh', 'fresh-menu').submenu('grm', 'grm-menu').submenu('kkt','kkt-menu').row()
  .back('←back');

// third level of menu const periodMenu = new Menu('period-menu', { autoAnswer: false })
  .text({text:'12 mth', payload:'12'}).text({text:'6 mth', payload: '6'}).row()
  .text({text:'3 mth', payload: '3'}).text({text:'1 mth', payload: '1'}).row()
  .text({text:'24 mth', payload: '24'}).back('←назад');

//registration
bot.use(menu);
menu.register(rateMenu);
rateMenu.register(periodMenu);
rateMenu.register(licenseMenu);
rateMenu.register(freshMenu);
rateMenu.register(grmMenu);
rateMenu.register(kktMenu);
// try to catch something
bot.on('callback_query', async (ctx) => {
  const [,,, payload] = ctx.callbackQuery.data.split('/')
  console.log(payload)
});

// log is like a period-menu/0/0/6/
// but I want to know data after each click of menu button

```

**Comment by MitkaZ7** [comment]
> hi, Can anybody help with this plugin? My menu with multiple submenus doesn't have any payloads. First level has 1-2 buttons, second level has multiple btns, and third level has final buttons, and i need to catch callback data at each level, but i catch it at only last level of menu? why? i can catch data without this plugin, via built-in InlineKeyboard only.
> 
> ```
>  // first level of menu
> const menu = new Menu('start-menu')
>   .submenu({ text: 'Price', payload: 'price'}, 'rate-menu');
> 
> // second level of menu
> export const rateMenu = new Menu('rate-menu')
>   .submenu({text:'prof', payload: 'prof'}, 'period-menu').submenu('base', 'period-menu')
>   .submenu('1C','license-menu').row()
>   .submenu('fresh', 'fresh-menu').submenu('grm', 'grm-menu').submenu('kkt','kkt-menu').row()
>   .back('←back');
> 
> // third level of menu const periodMenu = new Menu('period-menu', { autoAnswer: false })
>   .text({text:'12 mth', payload:'12'}).text({text:'6 mth', payload: '6'}).row()
>   .text({text:'3 mth', payload: '3'}).text({text:'1 mth', payload: '1'}).row()
>   .text({text:'24 mth', payload: '24'}).back('←назад');
> 
> //registration
> bot.use(menu);
> menu.register(rateMenu);
> rateMenu.register(periodMenu);
> rateMenu.register(licenseMenu);
> rateMenu.register(freshMenu);
> rateMenu.register(grmMenu);
> rateMenu.register(kktMenu);
> // try to catch something
> bot.on('callback_query', async (ctx) => {
>   const [,,, payload] = ctx.callbackQuery.data.split('/')
>   console.log(payload)
> });
> 
> // log is like a period-menu/0/0/6/
> // but I want to know data after each click of menu button
> ```

replace submenu() to text() and its work)
